File: 210-course-schedule-ii/course-schedule-ii.cs

public class Solution {
    public int[] FindOrder(int numCourses, int[][] prerequisites) {
        List<int>[] graph = new List<int>[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph[i] = new List<int>();
        }

        foreach (var prereq in prerequisites) {
            int course = prereq[0];
            int prerequisite = prereq[1];
            graph[prerequisite].Add(course);
        }

        Stack<int> stack = new Stack<int>();
        int[] visited = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            if (!DFS(i, graph, visited, stack)) {
                return new int[0]; 
            }
        }

        return stack.ToArray();
    }

    private bool DFS(int course, List<int>[] graph, int[] visited, Stack<int> stack) {
        if (visited[course] == 1) {
            return false; 
        }

        if (visited[course] == 2) {
            return true; 
        }

        visited[course] = 1;

        foreach (int nextCourse in graph[course]) {
            if (!DFS(nextCourse, graph, visited, stack)) {
                return false; 
            }
        }

        visited[course] = 2;
        stack.Push(course);

        return true;
    }
}

File: 210-course-schedule-ii/course-schedule-ii.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
    const inDegree=new Array(numCourses).fill(0);
    const graph=Array.from({length: numCourses},()=>[]);
    for (const [course, prerequisite] of prerequisites){
        inDegree[course]++;
        graph[prerequisite].push(course);
    }
    const queue=[];
    for (let i=0; i<numCourses; i++){
        if(inDegree[i]===0){queue.push(i);}
    }
    const order=[];
    while(queue.length >0){
        const current=queue.shift();
        order.push(current);
        for(const neighbor of graph[current]){
            inDegree[neighbor]--;
            if(inDegree[neighbor]===0){queue.push(neighbor);}
        }
    }
    if (order.length===numCourses){
        return order;
    }
    return [];
};